crates/noirc_frontend/src/hir_def/types.rs

```diff
@@ -894,4 +894,18 @@ impl Type {
             | Type::Unit => false,
         }
     }
+
+    pub fn flatten(&self) -> Vec<Type> {
+        let mut result = Vec::new();
+        match self {
+            Type::Struct(_, generic_args) => {
+                generic_args.iter().for_each(|t| result.extend(t.flatten()));
```

Comment:
Good point, I did not know that, although I wonder why it does not do a memcpy. I don't like to use append in general because it modifies the argument, but I believe in this case it is fine.

Reply:
`extend` cannot do a memcpy because it isn't specialized to `Vec`s, it only takes something it can iterate over. The most it can do is reserve capacity ahead of time before pushing one by one since some iterators have len methods.